control-plane/api-gateway/binding/route_binding.go

```diff
@@ -104,7 +104,22 @@ func (r *Binder) bindRoute(route client.Object, boundCount map[gwv1beta1.Section
 	for _, ref := range filteredParents {
 		var result bindResults
 
-		for _, listener := range listenersFor(&r.config.Gateway, ref.SectionName) {
+		listeners := listenersFor(&r.config.Gateway, ref.SectionName)
+
+		// If there are no matching listeners, then we failed to find the parent
+		if len(listeners) == 0 {
+			var sectionName gwv1beta1.SectionName
+			if ref.SectionName != nil {
+				sectionName = *ref.SectionName
+			}
+
+			result = append(result, bindResult{
+				section: sectionName,
+				err:     errRouteNoMatchingParent,
+			})
+		}
```

Comment:
This is part 1 of the core change where we detect the specific case while doing validation in order to construct the correct condition later on